pkg/queue/manager.go

```diff
@@ -547,3 +594,90 @@ func (m *Manager) reportPendingWorkloads(cqName string, cq ClusterQueue) {
 	}
 	metrics.ReportPendingWorkloads(cqName, active, inadmissible)
 }
+
+func (m *Manager) Start(ctx context.Context) error {
+	defer m.queue.ShutDown()
+
+	queueVisibilityUpdateInterval := time.Duration(m.queueVisibilityUpdateInterval) * time.Second
+
+	for i := 0; i < snapshotWorkers; i++ {
+		go wait.UntilWithContext(ctx, m.takeSnapshot, queueVisibilityUpdateInterval)
+	}
+
+	go wait.UntilWithContext(ctx, m.enqueueSnapshotDelayed, queueVisibilityUpdateInterval)
+
+	<-ctx.Done()
+
+	return nil
+}
+
+func (m *Manager) enqueueSnapshotDelayed(ctx context.Context) {
+	m.RLock()
+	defer m.RUnlock()
+	for cq := range m.clusterQueues {
+		m.queue.Add(cq)
+	}
+}
+
+func (m *Manager) takeSnapshot(ctx context.Context) {
+	for m.processNextSnapshot(ctx) {
+	}
+}
+
+func (m *Manager) processNextSnapshot(ctx context.Context) bool {
+	m.wm.Lock()
+	defer m.wm.Unlock()
+
+	log := ctrl.LoggerFrom(ctx).WithName("processNextSnapshot")
+
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+
+	startTime := time.Now()
+	defer func() {
+		log.V(4).Info("Finished snapshot job", "key", key, "elapsed", time.Since(startTime))
+	}()
+
+	cq := m.extractClusterQueue(key)
+	if cq == nil {
+		return false
+	}
+
+	defer m.queue.Done(key)
+
+	workloads := make([]kueue.ClusterQueuePendingWorkload, 0)
+	if elements, ok := cq.Snapshot(m.queueVisibilityClusterQueuesMaxCount); ok {
+		for _, wl := range elements {
+			workloads = append(workloads, kueue.ClusterQueuePendingWorkload{
+				Name:      wl.Name,
+				Namespace: wl.Namespace,
+			})
+		}
+	}
+	m.snapshots[key.(string)] = workloads
```

Comment:
IIUC t is better to have a function on the manager, say `SetSnapshot(cqName, workloads)` and take the lock there.